pkg/cmd/adm/restart.go

```diff
@@ -3,155 +3,198 @@ package adm
+			if err := checkRolloutStatus(f, ioStreams, "provider=codeready-toolchain"); err != nil {
+				return err
+			}
+		}
+	} else {
+		fmt.Printf("non-OLM based deployment not found in %s \n", ns)
+	}
 	return nil
 }
 
-func restartHostOperator(ctx *clicontext.CommandContext, hostClient runtimeclient.Client, hostNamespace string) error {
-	deployments := &appsv1.DeploymentList{}
-	if err := hostClient.List(context.TODO(), deployments,
-		runtimeclient.InNamespace(hostNamespace),
-		runtimeclient.MatchingLabels{"olm.owner.namespace": "toolchain-host-operator"}); err != nil {
+func deleteAndWaitForPods(ctx *clicontext.CommandContext, cl runtimeclient.Client, deployment appsv1.Deployment, f cmdutil.Factory, ioStreams genericclioptions.IOStreams) error {
+	fmt.Printf("Listing the pods to be deleted \n")
+	//get pods by label selector from the deployment
+	pods := corev1.PodList{}
+	selector, _ := metav1.LabelSelectorAsSelector(deployment.Spec.Selector)
+	if err := cl.List(ctx, &pods,
+		runtimeclient.MatchingLabelsSelector{Selector: selector},
+		runtimeclient.InNamespace(deployment.Namespace)); err != nil {
 		return err
 	}
-	if len(deployments.Items) != 1 {
-		return fmt.Errorf("there should be a single deployment matching the label olm.owner.namespace=toolchain-host-operator in %s ns, but %d was found. "+
-			"It's not possible to restart the Host Operator deployment", hostNamespace, len(deployments.Items))
+	fmt.Printf("Starting to delete the pods \n")
+	//delete pods
+	for _, pod := range pods.Items {
+		pod := pod // TODO We won't need it after upgrading to go 1.22: https://go.dev/blog/loopvar-preview
+		if err := cl.Delete(ctx, &pod); err != nil {
+			return err
+		}
+
+		fmt.Printf("Checking the status of the rolled out deployment %v", deployment)
+		//check the rollout status
+		if err := checkRolloutStatus(f, ioStreams, "kubesaw-control-plane=kubesaw-controller-manager"); err != nil {
+			return err
+		}
+	}
+	return nil
+
+}
+
+func restartNonOlmDeployments(deployment appsv1.Deployment, f cmdutil.Factory, ioStreams genericclioptions.IOStreams) error {
+
+	o := kubectlrollout.NewRolloutRestartOptions(ioStreams)
+
+	if err := o.Complete(f, nil, []string{"deployment"}); err != nil {
+		panic(err)
 	}
 
-	return restartDeployment(ctx, hostClient, hostNamespace, deployments.Items[0].Name)
+	o.Resources = []string{"deployment/" + deployment.Name}
+
+	if err := o.Validate(); err != nil {
+		panic(err)
+	}
+	fmt.Printf("Running the rollout restart command for non-olm deployment %v", deployment)
+	return o.RunRestart()
 }
 
-func printExistingDeployments(term ioutils.Terminal, cl runtimeclient.Client, ns string) error {
-	deployments := &appsv1.DeploymentList{}
-	if err := cl.List(context.TODO(), deployments, runtimeclient.InNamespace(ns)); err != nil {
-		return err
+func checkRolloutStatus(f cmdutil.Factory, ioStreams genericclioptions.IOStreams, labelSelector string) error {
```

Comment:
this is fine for checking the latest rollout status, but it doesn't ensure that we actually have new pods right? 
Should we introduce a check and verify that there are newly created pods after the restart ?

Reply:
This may be useful with the OLM based deployments since there we directly delete the pods thus we want to make sure new ones are created.